suite/tests/api/detach_state.c

```
@@ -410,6 +410,111 @@ test_thread_func(void (*asm_func)(void))
+     */
+    if ((ptr_uint_t)tag % 2 == 0)
+        return DR_EMIT_DEFAULT;
+    else
+        return DR_EMIT_STORE_TRANSLATIONS;
+}
+
+static bool
+event_restore(void *drcontext, bool restore_memory, dr_restore_state_info_t *info)
+{
+#    ifdef X86
+    /* Because we set the translation (to avoid detach timeouts) we need to restore
+     * our register too.
+     */
+    if (info->fragment_info.cache_start_pc == NULL ||
+        /* At the first instr should require no translation. */
+        info->raw_mcontext->pc <= info->fragment_info.cache_start_pc)
+        return true;
+    instr_t inst;
+    instr_init(drcontext, &inst);
+    byte *pc = info->fragment_info.cache_start_pc;
+    while (pc < info->raw_mcontext->pc) {
+        instr_reset(drcontext, &inst);
+        pc = decode(drcontext, pc, &inst);
+        bool tls;
+        uint offs;
+        bool spill;
+        reg_id_t reg;
+        if (instr_is_reg_spill_or_restore(drcontext, &inst, &tls, &spill, &reg, &offs) &&
+            tls && !spill && reg == DR_REG_XAX) {
+            /* The target point is past our restore. */
+            instr_free(drcontext, &inst);
+            return true;
+        }
+    }
+    instr_free(drcontext, &inst);
+    reg_set_value(DR_REG_XAX, info->mcontext, dr_read_saved_reg(drcontext, SPILL_SLOT_1));
+#    endif
+    return true;
+}
+
+static void
+event_exit(void)
+{
+    dr_custom_free(NULL, 0, load_from, dr_page_size());
+    dr_unregister_bb_event(event_bb);
+    dr_unregister_restore_state_ex_event(event_restore);
+    dr_unregister_exit_event(event_exit);
+}
+
+DR_EXPORT void
+dr_client_main(client_id_t id, int argc, const char *argv[])
+{
+    static bool do_once;
+    if (!do_once) {
+        print("in dr_client_main\n");
+        do_once = true;
+    }
+    dr_register_bb_event(event_bb);
+    dr_register_restore_state_ex_event(event_restore);
+    dr_register_exit_event(event_exit);
+
+    /* Try to get an address that is not 32-bit-displacement reachable from
```

ASSERT that it's indeed not reachable to avoid the test silently losing utility.

---

This is not simple to do -- the best I found was when we have a cache address in event_restore().  Done.